pandas/core/arrays/categorical.py

```diff
@@ -1229,11 +1229,26 @@ def map(self, mapper):
         Index(['first', 'second', nan], dtype='object')
         """
         new_categories = self.categories.map(mapper)
+
         try:
-            return self.from_codes(self._codes.copy(),
-                                   categories=new_categories,
-                                   ordered=self.ordered)
+            if is_dict_like(mapper):
+                new_value = mapper[np.nan]
+            else:
+                new_value = mapper(np.nan)
+        except (AttributeError, KeyError, TypeError, ValueError):
```

Comment:
I'm not really comfortable with this. `mapper` is a user-defined function. Consider

```python
def f(x):
    if isnan(x):
        raise TypeError
    ...
```

that TypeError would be swallowed by pandas.